test/test.c

```diff
@@ -28636,6 +28639,495 @@ void test_JetStreamSubscribePull(void)
+        s = js_PullSubscribeAsync(NULL, js, "", "dur", _recvPullAsync, &args, NULL, NULL, NULL, &jerr);
+    if (s == NATS_INVALID_ARG)
+        s = js_PullSubscribeAsync(&sub, js, "foo", "dur", NULL, &args, NULL, NULL, NULL, &jerr);
+    testCond((s == NATS_INVALID_ARG) && (sub == NULL) && (jerr == 0));
+    nats_clearLastError();
+
+    test("AckNone ok: ");
+    jsSubOptions_Init(&so);
+    so.Config.AckPolicy = js_AckNone;
+    s = js_PullSubscribeAsync(&sub, js, "foo", "ackNone", _recvPullAsync, &args, NULL, NULL, &so, &jerr);
+    testCond((s == NATS_OK) && (sub != NULL) && (jerr == 0));
+    natsSubscription_Unsubscribe(sub);
+    natsSubscription_Destroy(sub);
+    sub = NULL;
+
+    test("AckAll ok: ");
+    jsSubOptions_Init(&so);
+    so.Config.AckPolicy = js_AckAll;
+    s = js_PullSubscribeAsync(&sub, js, "foo", "ackAll", _recvPullAsync, &args, NULL, NULL, &so, &jerr);
+    testCond((s == NATS_OK) && (sub != NULL) && (jerr == 0));
+    natsSubscription_Unsubscribe(sub);
+    natsSubscription_Destroy(sub);
+    sub = NULL;
+
+    test("Create push consumer: ");
+    jsConsumerConfig_Init(&cc);
+    cc.Durable = "push_dur";
+    cc.DeliverSubject = "push.deliver";
+    s = js_AddConsumer(NULL, js, "TEST", &cc, NULL, &jerr);
+    testCond((s == NATS_OK) && (jerr == 0));
+
+    test("Try create pull sub from push consumer: ");
+    s = js_PullSubscribeAsync(&sub, js, "foo", "push_dur", _recvPullAsync, &args, NULL, NULL, &so, &jerr);
+    testCond((s == NATS_ERR) && (sub == NULL) && (jerr == 0) && (strstr(nats_GetLastError(NULL), jsErrPullSubscribeToPushConsumer) != NULL));
+    nats_clearLastError();
+
+    test("Create pull bound failure: ");
+    jsSubOptions_Init(&so);
+    so.Stream = "TEST";
+    so.Consumer = "bar";
+    s = js_PullSubscribeAsync(&sub, js, "foo", "bar", _recvPullAsync, &args, NULL, NULL, &so, &jerr);
+    testCond((s == NATS_NOT_FOUND) && (sub == NULL) && (jerr == JSConsumerNotFoundErr));
+    nats_clearLastError();
+
+    // TEST real subscription, auto-ack.
+
+    test("js_PullAsyncSubscribe for real, manual ack: ");
+    // Signal as soon as we get the first message.
+    args.control = 2; // don't ack
+
+    jsSubOptions_Init(&so);
+    so.Config.MaxAckPending = 10;
+    so.Config.AckWait = NATS_MILLIS_TO_NANOS(300);
+    so.ManualAck = true;
+    s = js_PullSubscribeAsync(&sub, js, "foo", "dur", _recvPullAsync, &args, NULL, NULL, &so, &jerr);
+    testCond((s == NATS_OK) && (sub != NULL) && (jerr == 0));
+
+    test("Can't call NextMsg: ");
+    s = natsSubscription_NextMsg(&msg, sub, 1000);
+    testCond((s == NATS_ILLEGAL_STATE) && (msg == NULL));
+    nats_clearLastError();
+
+    test("Can't call Fetch: ");
+    s = natsSubscription_Fetch(&list, sub, 1, 500, &jerr);
+    testCond((s == NATS_ERR) && (msg == NULL) && (strstr(nats_GetLastError(NULL), jsErrConcurrentFetchNotAllowed) != NULL));
+    nats_clearLastError();
+
+    int noMessageTimeout = 40;
+    int messageArrivesImmediatelyTimeout = 20;
```

Comment:
On my Windows VM, this test flaps a lot, and I wonder if you should not bump (either for all cases or just under windows) those 2 values,  maybe double: 80/40?

Reply:
Updated, let's see?

Reply:
Even on macOS, with repeat of tests, I get:
```
== JetStreamSubscribePullAsync ==
#01 Start JS Server: PASSED
#02 Connect: PASSED
#03 Get context: PASSED
#04 Create stream: PASSED
#05 Create pull sub async (invalid args): PASSED
#06 AckNone ok: PASSED
#07 AckAll ok: PASSED
#08 Create push consumer: PASSED
#09 Try create pull sub from push consumer: PASSED
#10 Create pull bound failure: PASSED
#11 js_PullAsyncSubscribe for real, manual ack: PASSED
#12 Can't call NextMsg: PASSED
#13 Can't call Fetch: PASSED
#14 No messages yet, timeout in 80ms: PASSED
#15 Send a message: PASSED
#16 Arrives in under 40ms: PASSED
#17 No more messages yet, timeout in 80ms: PASSED
#18 Wait another so.Config.AckWait+100 = 400ms and see that the message is re-delivered, ACK it: PASSED
#19 Wait another so.Config.AckWait+100 = 400ms and see that the message is not re-delivered: PASSED
#20 Receive msg with header and no data comes through: PASSED
#21 single fetch NOWAIT, partially fulfilled NATS_TIMEOUT: PASSED
#22 multi-fetch NOWAIT, partially fulfilled NATS_TIMEOUT: PASSED
#23 fetch NOWAIT can get a NATS_NOT_FOUND immediately: PASSED
#24 fetch NOWAIT can get a NATS_NOT_FOUND on Nth request: PASSED
#25 single fetch NOWAIT, fulfilled msgs NATS_MAX_DELIVERED_MSGS: PASSED
#26 multi-fetch NOWAIT, fulfilled msgs NATS_MAX_DELIVERED_MSGS: PASSED
#27 single fetch, fulfilled msgs NATS_MAX_DELIVERED_MSGS: PASSED
#28 multi-fetch WAIT, fulfilled msgs NATS_MAX_DELIVERED_MSGS: PASSED
#29 MaxBytes: PASSED
#30 Fetch with expiration is fulfilled NATS_MAX_DELIVERED_MSGS: PASSED
#31 Fetch with a short expiration is partially fulfilled NATS_TIMEOUT: FAILED: status: 22, expected: 26
FAILED
*** TEST FAILED ***
```
or on Windows after the first run:
```
== JetStreamSubscribePullAsync ==
#01 Start JS Server: PASSED
#02 Connect: PASSED
#03 Get context: PASSED
#04 Create stream: PASSED
#05 Create pull sub async (invalid args): PASSED
#06 AckNone ok: PASSED
#07 AckAll ok: PASSED
#08 Create push consumer: PASSED
#09 Try create pull sub from push consumer: PASSED
#10 Create pull bound failure: PASSED
#11 js_PullAsyncSubscribe for real, manual ack: PASSED
#12 Can't call NextMsg: PASSED
#13 Can't call Fetch: PASSED
#14 No messages yet, timeout in 80ms: PASSED
#15 Send a message: PASSED
#16 Arrives in under 40ms: PASSED
#17 No more messages yet, timeout in 80ms: PASSED
#18 Wait another so.Config.AckWait+100 = 400ms and see that the message is re-delivered, ACK it: PASSED
#19 Wait another so.Config.AckWait+100 = 400ms and see that the message is not re-delivered: PASSED
#20 Receive msg with header and no data comes through: PASSED
#21 single fetch NOWAIT, partially fulfilled NATS_TIMEOUT: PASSED
#22 multi-fetch NOWAIT, partially fulfilled NATS_TIMEOUT: PASSED
#23 fetch NOWAIT can get a NATS_NOT_FOUND immediately: PASSED
#24 fetch NOWAIT can get a NATS_NOT_FOUND on Nth request: PASSED
#25 single fetch NOWAIT, fulfilled msgs NATS_MAX_DELIVERED_MSGS: PASSED
#26 multi-fetch NOWAIT, fulfilled msgs NATS_MAX_DELIVERED_MSGS: FAILED: timed out: 26
FAILED: onComplete has not been called
FAILED: status: 0, expected: 22
FAILED: msgs: 75, expected: 100
FAILED: subscription is still valid
FAILED
*** TEST FAILED ***
```

Reply:
Resolved in favor of thread https://github.com/nats-io/nats.c/pull/785#discussion_r1725839401